### PR TITLE
Add key mapping and Update CameraController using improved Input system.

### PR DESCRIPTION
**Pull request by 3x3y3z3t**
#### Describe the issue (if no issue has been made)
This PR doesn't fix any issue.
This PR will improve the existing `Input` and `OrthographicCameraController` described below.

#### PR impact _(Make sure to add [closing keywords](https://help.github.com/en/articles/closing-issues-using-keywords))_
List of related issues/PRs this will solve:

 Impact                  | Issue/PR
------------------------ | ------
Other PRs this solves    | #161, #124  

#### Proposed fix _(Make sure you've read [on how to contribute](https://github.com/TheCherno/Hazel/blob/master/.github/CONTRIBUTING.md) to Hazel)_
This PR adds/improves the following:
- KeyCodes/MouseButtonCodes:
  - Add `enum class` `Keys` and `MouseButtons` (#124).
  - In the future version, macros will be replaced completely by those enum classes.
- Input: 
  - Add `KeyAlternative` struct which represent a single key press or mouse button press.
  - Input checks now checks over the said `KeyAlternative` instead of raw keycodes/mousebuttoncodes `int` values.
  - Add a map for key mapping: each `std::string` key is assigned with one `KeyAlternative`, and are remappable at any time/any place.
- OrthographicCameraController:
  - Update Controller to take advantage of the new Keymapping system.
  - **[minor]** Zoom Level and Enable Rotation can now be manually set at any time/any place. There is a min/max zoom level limitation.
  - **[minor]** Camera translation speed now scales with the zoom level (#161).

This PR is backward-compatible with current version of `master` branch Hazel (need minor variable rename though).

#### Additional context
None.

### CODE REVIEW

File: Hazel/src/Hazel/Renderer/OrthographicCameraController.cpp [hunk 1]
```diff
@@ -62,18 +75,34 @@ namespace Hazel {
...
+	void OrthographicCameraController::SetZoomLevel(float level) 
+	{
+		// translation speed should scale relative to the zoom level changes;
+		float diff = level / m_ZoomLevel;
+		m_CameraTranslationSpeed *= diff;
+		m_ZoomLevel = level;
+		UpdateProjectionMatrix();
+	}
+
+	void OrthographicCameraController::UpdateProjectionMatrix()
```
Comment (lhurt51) [inline Hazel/src/Hazel/Renderer/OrthographicCameraController.cpp]: This function could be an inline but all in all great work I like the idea of function mapping like many of the big game engines have!

File: Hazel/src/Hazel/Renderer/OrthographicCameraController.cpp [hunk 2]
```diff
@@ -62,18 +75,34 @@ namespace Hazel {
...
-		m_ZoomLevel -= e.GetYOffset() * 0.25f;
-		m_ZoomLevel = std::max(m_ZoomLevel, 0.25f);
 		m_Camera.SetProjection(-m_AspectRatio * m_ZoomLevel, m_AspectRatio * m_ZoomLevel, -m_ZoomLevel, m_ZoomLevel);
+	}
+
+	bool OrthographicCameraController::OnMouseScrolled(MouseScrolledEvent& e)
+	{
+		float zoomLevel = m_ZoomLevel - e.GetYOffset() * m_ZoomRate;
+		zoomLevel = std::max(zoomLevel, m_MaxZoomLevel);
+		zoomLevel = std::min(zoomLevel, m_MinZoomLevel);
```
Comment (Emilundpixeln) [inline Hazel/src/Hazel/Renderer/OrthographicCameraController.cpp]: Maybe just use `std::clamp` instead of `std::min` and `std::max`.

File: Hazel/src/Hazel/Core/KeyCodes.h [hunk 3]
```diff
@@ -1,5 +1,142 @@
 #pragma once
 
+namespace Hazel
+{
+    enum class Keys
```
Comment (LovelySanta) [inline Hazel/src/Hazel/Core/KeyCodes.h]: When using this enum, we write `Keys::A`. Maybe create a typedef to map Keys on Key so we can write `Key::A`?
```cpp
typedef enum class Keys
{
    ... // key codes
} Key;
```
or
```cpp
enum class Keys
{
    ... // key codes
};
using Key = Keys;
```

File: Hazel/src/Hazel/Core/MouseButtonCodes.h [hunk 4]
```diff
@@ -1,6 +1,32 @@
 #pragma once
 
-// From glfw3.h
+namespace Hazel
+{
+    enum class MouseButtons
```
Comment (LovelySanta) [inline Hazel/src/Hazel/Core/MouseButtonCodes.h]: Same comments as with the Keys:
- Convert spaces to tabs
- Definition of the base type `int16_t` (?)
```suggestion
    enum class MouseButtons : int16_t
```

Comment (LovelySanta) [inline Hazel/src/Hazel/Core/MouseButtonCodes.h]: Maybe rename this so `MouseButtons::Button1` becomes `Mouse::Button1` to be inline with the key enum.
```suggestion
    enum class Mouse
```

_Here there is no typedef required as we only use 1 mouse._ ;)

File: Hazel/src/Hazel/Core/Input.h [hunk 5]
```diff
@@ -12,12 +28,41 @@ namespace Hazel {
...
+		inline static bool IsKeyPressed(Keys key) { return s_Instance->IsKeyPressedImpl((int)key); }
+		/* This function is still kept to ensure backward-compatibility. */
 		inline static bool IsMouseButtonPressed(int button) { return s_Instance->IsMouseButtonPressedImpl(button); }
+		inline static bool IsMouseButtonPressed(MouseButtons button) { return s_Instance->IsMouseButtonPressedImpl((int)button); }
+
 		inline static std::pair<float, float> GetMousePosition() { return s_Instance->GetMousePositionImpl(); }
 		inline static float GetMouseX() { return s_Instance->GetMouseXImpl(); }
 		inline static float GetMouseY() { return s_Instance->GetMouseYImpl(); }
+
+		inline static void Remap(const std::string& name, KeyAlternative input) { s_Instance->m_Mapping[name] = input; }
```
Comment (LovelySanta) [inline Hazel/src/Hazel/Core/Input.h]: This remap feature (and use in the input class) should move to it's own PR. It can implement #161 with it.

**Comment by 3x3y3z3t** [comment]
Sorry for the long delay (I'm moving and my new place didn't have internet connection set up).
I have made some changes as suggested by @LovelySanta in my local code.
Since @IshoAntar have updated his PR #124, this PR become a duplicate of his one. Should I close this PR and create a new one for key mapping, or should I just strip out the new Keycode/Mousecode part?

File: Hazel/src/Hazel/Core/Input.h [hunk 6]
```diff
@@ -1,9 +1,25 @@
...
+#include "Hazel/Core/MouseButtonCodes.h"
+
+#include "Hazel/Core/Log.h"
 
 namespace Hazel {
 
+	/* This struct holds a single input, either a key code or a mouse button code.
+	 * Currently multiple alternative input is not supported.
+	 */
+	struct KeyAlternative
```
[the remark this answers is not in the document]
Comment (3x3y3z3t) [reply Hazel/src/Hazel/Core/Input.h]: My idea for the `KeyAlternative` struct is so we don't have to hold a two separated map for `Keys` and `MouseButtons`.
When user map an input, for example 'Jump', some may want to use left mouse button, some use space key. Whatever the user will map (key or mouse), the map will only see it as an "Input key". When the user want to get that 'Jump' input, the map will return that "Input key".

[the remark this answers is not in the document]
Comment (LovelySanta) [reply Hazel/src/Hazel/Core/Input.h]: > When user map an input, for example 'Jump', some may want to use left mouse button, some use space key. Whatever the user will map (key or mouse), the map will only see it as an "Input key". When the user want to get that 'Jump' input, the map will return that "Input key".

I did not think of that indeed, which is a valid point.

**Comment by LovelySanta** [comment]
> Sorry for the long delay (I'm moving and my new place didn't have internet connection set up).
> I have made some changes as suggested by @LovelySanta in my local code.
> Since @IshoAntar have updated his PR #124, this PR become a duplicate of his one. Should I close this PR and create a new one for key mapping, or should I just strip out the new Keycode/Mousecode part?

I would suggest leave this one open untill you have your new PR up. Also, maybe wait from making your PR until #124 is merged (will happen with my next merge testing, when I have time)

**Comment by 3x3y3z3t** [comment]
> I would suggest leave this one open untill you have your new PR up. Also, maybe wait from making your PR until #124 is merged (will happen with my next merge testing, when I have time)
Thanks.
Since the new input works with `Keys` and `Mouse` enum class instead of raw integer values, I will have to wait until #124 is merged and those enum classes are part of the main code base, or else it will become a broken PR :))
